### ISSUE TITLE
Support pre-suspend and post-resume hooks

### ISSUE THREAD

**Issue by arbrandes** [hastexo/hastexo-xblock#97]
## User story

As a course author, I can write a POSIX script (bash or otherwise) and include it in the courseware in such a way that it can be automatically invoked via SSH on the VM accessible via the stack's floating IP immediately prior to a stack suspend event with a "suspend" argument, immediately prior to a stack delete with "delete" argument, and immediately following a stack resume event with a "resume" argument.  I can also select which of these events I want the script to hook to.

## Use cases

The primary use case is to work around a KVM limitation that forbids the managed save of a host VM if it has _running_ nested KVM instances.  Such a script would allow the XBlock to run a `virsh managedsave` on any nested VMs prior to a `stack suspend`, and a subsequent `virsh start` on them after resume, thus circumventing the KVM limitation.  This is an example of such a script:

```
#!/bin/bash -e
#/tmp/suspend_hook.sh

ACTION=${1}

if [ -z "$ACTION" ]
then
    echo "Please specify an action, one of 'suspend' or 'resume'." >&2
    exit 1
fi

if [ "$ACTION" = "suspend" ]
then
    for domain in $(virsh list --name)
    do
        virsh managedsave $domain
    done
elif [ "$ACTION" = "resume" ]
then
    for domain in $(virsh list --name --all)
    do
        if virsh dominfo $domain | grep -q 'Managed\ save:.*yes'
        then
            virsh start $domain
        fi
    done
else
    echo "Unsupported action '$ACTION'." >&2
    exit 1
fi
```

There is one other foreseen use case, which is cleaning up external resources (such as Magnum stacks) prior to the stack being deleted.

### Definition of the hook script

Much like the lab template definitions, the course author will upload a file with an arbitrary name to the content store.  They will then specify a path to the file via a global `hook_script` XBlock string field.

### Selection of events

The course author will define a `hook_events` XBlock dict field, where each entry is a (string, boolean) value pair, the first being one of "suspend", "resume", "delete", the second being "True" if they desire the script to be run on that event, "False" if not.

If a dict entry is omitted from `hook_events` for any event, it will be considered to be "False".  In other words, the default behavior for any event is to _not_ run the hook script.

### Execution of pre-suspend hook

The `suspender` command will be modified to execute the script at the appropriate time.  Some considerations:

* The script has access to the stack name and SSH key via the `hastexo_stack` table.
* The path to the hook script will be saved to a string field in the `hastexo_stack` table.  The suspender job will then fetch it from the content store.
* The hook event selection will be saved to a JSONField in the `hastexo_stack` table.
* Until now, to minimize the chance of holding up the suspend queue, suspender execution per stack would be a quick affair.  With this proposal, where the pre-suspend script execution could take minutes, each stack suspend will hang around much longer.  A way will have to be found for allowing new suspender runs to happen on intervals shorter than what it takes to suspend one stack, as currently the APS scheduler requires a run to be complete before starting the next one.  Allowing x simultaneous runs might be an option, but the implementation would have to make sure to avoid database deadlocks and other concurrency issues.
* Since the suspender will now hang around longer anyway, it should be modified so that it waits for successful stack suspension, and retry in case of failure.

### Execution of post-resume hook

This one is simpler than the pre-suspend hook, as the launch Celery task already has access to all the necessary information, and is also expected to be long-running.  All that's needed is the addition of a step immediately after SSH verification.

### Execution of pre-delete hook

Since by definition stacks will be suspended prior to a delete event, if it is selected by the author, the stack will be resumed prior to execution, after which it will be deleted.  Otherwise, like the pre-suspend hook, it will fetch the script from the content store, and whether or not to execute the delete script from the database.

### Other considerations

Since the `reboot_on_resume` template output was only ever used to work around the same KVM limitation being dealt with here (albeit in a less-than-ideal manner), it will be deprecated to be removed on the following major release.

**Comment by fghaas** [comment]
Overall, I like the proposal. I do have a few suggestions though:

1. I contemplated for a bit if we should allow the hook scripts to be in the target VM, as opposed to the content store. I now think, though, that this is not such a great idea, as this would allow an unsuspecting learner to delete the script from the VM's filesystem and thus inadvertently break their suspend/resume capability.

2. I think both the stack output and the database fields should be more generically named, as we might want other hooks in the future. In particular, one thing I have wanted for a long time is a pre-delete hook, i.e. something that does cleanup jobs before a stack is reaped. I can think of a total of four hooks that might be useful (post-create, pre-suspend, post-resume, and pre-delete), although there might be others. Post-create is arguably already covered by cloud-init, but all cloud-init data is always accessible to the learner, and we might want to run some post-creation tasks we don't want to expose to learners.

3. `suspend_script_servers` won't work, with the current proposal, if there is more than one `OS::Nova::Server` instance needing special care and using IPv4, but the stack has only one floating IP. The XBlock would only ever be able to connect to the one floating IP, and wouldn't be able to SSH into other Nova instances directly. I thus propose to not deal with server UUIDs anymore like we presently do the the post-suspend reboot, but instead to use IP addresses.

   Ideally, I'd suggest to specify a Heat output named `hooks`, as a dictionary, such as:

   ```yaml
   outputs:
     public_ip:
       description: Floating IP address in public network
       value: { get_attr: [ server1, floating_ip_address ] }
     hooks:
       description: Hooks to run for specific events
       value:
         # suppose only server2 needs special care for suspend/resume,
         # and only server1 needs to run a pre-delete hook 
         post_create:
           - { get_attr: [ server1, floating_ip_address ] }
           - { get_attr: [ server2, floating_ip_address ] }
         pre_suspend:
           - { get_attr: [ server2, floating_ip_address ] }
         post_resume:
           - { get_attr: [ server2, floating_ip_address ] }
         pre_delete:
           - { get_attr: [ server1, floating_ip_address ] }
   ```

   Then, the XBlock retrieve the `hooks` output as a JSON dictionary, parse it, cycle through all required servers, and subsequently suspend the stack. If the servers are accessible via IPv6, of course, then the hooks dictionary would refer to the servers by those addresses, rather than by floating IP.

4. I believe if we do take up this new SSH-based approach as opposed to Nova reboots, we deprecate, and in a later release remove, *all* dependencies on python-novaclient, so that we cleanly interact with only Heat (or Google Cloud Deployment Manager) anymore.

**Comment by arbrandes** [comment]
>     1. I contemplated for a bit if we should allow the hook scripts to be in the target VM, as opposed to the content store. I now think, though, that this is not such a great idea, as this would allow an unsuspecting learner to delete the script from the VM's filesystem and thus inadvertently break their suspend/resume capability.

I considered it, and rejected it, as well.  This is the same reason why we don't include progress checks in the VM.

>     2. I think both the stack output and the database fields should be more generically named, as we might want other hooks in the future. In particular, one thing I have wanted for a long time is a pre-delete hook, i.e. something that does cleanup jobs before a stack is reaped. I can think of a total of four hooks that might be useful (post-create, pre-suspend, post-resume, and pre-delete), although there might be others. Post-create is arguably already covered by cloud-init, but all cloud-init data is always accessible to the learner, and we might want to run some post-creation tasks we don't want to expose to learners.

Good idea.  Agreed.  This means that the reaper job also gets a hook execution step.  And that case is almost as simple as the post-create/post-resume one, as both launch task and delete job are expected to take a while to complete.

>     3. `suspend_script_servers` won't work, with the current proposal, if there is more than one `OS::Nova::Server` instance needing special care and using IPv4, but the stack has only one floating IP. The XBlock would only ever be able to connect to the one floating IP, and wouldn't be able to SSH into other Nova instances directly. I thus propose to not deal with server UUIDs anymore like we presently do the the post-suspend reboot, but instead to use IP addresses.

Nice catch.  It somehow slipped my mind that UUIDs would be no good for SSHing in, for obvious reasons.

>        Ideally, I'd suggest to specify a Heat output named `hooks`, as a dictionary, such as:
>        ```yaml
>        outputs:
>          public_ip:
>            description: Floating IP address in public network
>            value: { get_attr: [ server1, floating_ip_address ] }
>          hooks:
>            description: Hooks to run for specific events
>            value:
>              # suppose only server2 needs special care for suspend/resume,
>              # and only server1 needs to run a pre-delete hook 
>              post_create:
>                - { get_attr: [ server1, floating_ip_address ] }
>                - { get_attr: [ server2, floating_ip_address ] }
>              pre_suspend:
>                - { get_attr: [ server2, floating_ip_address ] }
>              post_resume:
>                - { get_attr: [ server2, floating_ip_address ] }
>              pre_delete:
>                - { get_attr: [ server1, floating_ip_address ] }
>        ```
>        Then, the XBlock retrieve the `hooks` output as a JSON dictionary, parse it, cycle through all required servers, and subsequently suspend the stack. If the servers are accessible via IPv6, of course, then the hooks dictionary would refer to the servers by those addresses, rather than by floating IP.

For stacks with a single IPv4 floating IP and multiple VMs, the floating IP alone won't be enough.  We also need the internal IP or host name.  And copying a tmp script and running it inception-style sounds like a Paramiko headache.  Plus, we've managed to avoid having the XBlock have to discover stack internals as much as we could, so far, as it simplifies template requirements.  

So... How about we do exactly as the progress checks (which suffer the same problems), and leave this entirely up to the script?  Since internal VM IPs are as predictable as the course author makes them anyway, it gives the hooks maximum flexibility to do anything they please with the _entire_ environment.  _And_ it makes the XBlock code that much simpler.

In other words, it's up to the course author to set up the VMs so that a hook script run on a bastion node does what it needs to with the whole environment.  Want to have an Ansible playbook in the VM that gets called pre-suspend?  Sure, why not?  Prefer a Salt agent?  Or just SSHing into multiple VMs from the script?  Go on right ahead!

This would mean, then, that we simply deprecate the `reboot_on_resume` and its Nova dependency, instead of hijacking it.

>     4. I believe if we do take up this new SSH-based approach as opposed to Nova reboots, we deprecate, and in a later release remove, _all_ dependencies on python-novaclient, so that we cleanly interact with only Heat (or Google Cloud Deployment Manager) anymore.

Correct!

**Comment by arbrandes** [comment]
There's a hitch with the pre-delete hook: by definition, the stack will be suspended by the time the reaper acts.  What we can do is if there's delete hook defined, resume the stack, run the hook, and _then_ delete it.  Objections?

**Comment by fghaas** [comment]
> So... How about we do exactly as the progress checks (which suffer the same problems), and leave this entirely up to the script? Since internal VM IPs are as predictable as the course author makes them anyway, it gives the hooks maximum flexibility to do anything they please with the _entire_ environment. _And_ it makes the XBlock code that much simpler.

So you're saying we are not even adding any additional outputs, but instead just rely on the one `public_ip` that we have?

> There's a hitch with the pre-delete hook: by definition, the stack will be suspended by the time the reaper acts. What we can do is if there's delete hook defined, resume the stack, run the hook, and _then_ delete it. Objections?

None.

**Comment by arbrandes** [comment]
> So you're saying we are not even adding any additional outputs, but instead just rely on the one `public_ip` that we have?

Correct.

And if the time comes when there's no alternative but to give the XBlock direct access to internal VMs, we can then think about building a general framework for it, instead of just for hook scripts.  But as I wrote earlier, this will have a not insignificant drawback of placing yet more general restrictions on lab templates, so if we can avoid it, so much the better.

**Comment by fghaas** [comment]
>>So you're saying we are not even adding any additional outputs, but instead just rely on the one public_ip that we have?

> Correct.

> And if the time comes when there's no alternative but to give the XBlock direct access to internal VMs, we can then think about building a general framework for it, instead of just for hook scripts. But as I wrote earlier, this will have a not insignificant drawback of placing yet more general restrictions on lab templates, so if we can avoid it, so much the better.

Ack on both.

**Comment by fghaas** [comment]
> Ack on both.

Hang on, that's not how the cool kids do it these days. I have added appropriate emoji reactions.

**Comment by arbrandes** [comment]
Ok, I've updated the issue description with our findings.

**Comment by fghaas** [comment]
> The course author will define a hook_events XBlock dict field, where each entry is a (string, boolean) value pair, the first being one of "suspend", "resume", "delete", the second being "True" if they desire the script to be run on that event, "False" if not.

Please add that if any of the known keys is omitted from the dictionary, it will be considered equivalent to False. We don't want XBlock failures because a course author forgot to define one key.

**Comment by arbrandes** [comment]
Good call.  Updated!

**Comment by fghaas** [comment]
Implemented in #101 (468e31287043b9bac783a1b9f08d25abef156c22).
